src/main/java/org/opensearch/security/support/Base64JDKHelper.java

```diff
@@ -0,0 +1,199 @@
+        Pattern.class,
+        User.class,
+        SourceFieldsContext.class,
+        LdapUser.class,
+        SearchEntry.class,
+        LdapEntry.class,
+        AbstractLdapBean.class,
+        LdapAttribute.class
+    );
+
+    private static final List<Class<?>> SAFE_ASSIGNABLE_FROM_CLASSES = ImmutableList.of(
+        InetAddress.class,
+        Number.class,
+        Collection.class,
+        Map.class,
+        Enum.class
+    );
+
+    private static final Set<String> SAFE_CLASS_NAMES = Collections.singleton("org.ldaptive.LdapAttribute$LdapAttributeValues");
+
+    private static boolean isSafeClass(Class<?> cls) {
+        return cls.isArray()
+            || SAFE_CLASSES.contains(cls)
+            || SAFE_CLASS_NAMES.contains(cls.getName())
+            || SAFE_ASSIGNABLE_FROM_CLASSES.stream().anyMatch(c -> c.isAssignableFrom(cls));
+    }
+
+    private final static class SafeObjectOutputStream extends ObjectOutputStream {
+
+        private static final boolean useSafeObjectOutputStream = checkSubstitutionPermission();
+
+        @SuppressWarnings("removal")
+        private static boolean checkSubstitutionPermission() {
+            SecurityManager sm = System.getSecurityManager();
+            if (sm != null) {
+                try {
+                    sm.checkPermission(new SpecialPermission());
+
+                    AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                        AccessController.checkPermission(SUBSTITUTION_PERMISSION);
+                        return null;
+                    });
+                } catch (SecurityException e) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        static ObjectOutputStream create(ByteArrayOutputStream out) throws IOException {
+            try {
+                return useSafeObjectOutputStream ? new SafeObjectOutputStream(out) : new ObjectOutputStream(out);
+            } catch (SecurityException e) {
+                // As we try to create SafeObjectOutputStream only when necessary permissions are granted, we should
+                // not reach here, but if we do, we can still return ObjectOutputStream after resetting ByteArrayOutputStream
+                out.reset();
+                return new ObjectOutputStream(out);
+            }
+        }
+
+        @SuppressWarnings("removal")
+        private SafeObjectOutputStream(OutputStream out) throws IOException {
+            super(out);
+
+            SecurityManager sm = System.getSecurityManager();
+            if (sm != null) {
+                sm.checkPermission(new SpecialPermission());
+            }
+
+            AccessController.doPrivileged((PrivilegedAction<Boolean>) () -> enableReplaceObject(true));
+        }
+
+        @Override
+        protected Object replaceObject(Object obj) throws IOException {
+            Class<?> clazz = obj.getClass();
+            if (isSafeClass(clazz)) {
```

Comment:
Don't we need similar checks with custom serialization as well?

Reply:
We won't need the safe class checks for Custom Serialization as we explicitly configure each of the classes eligible for (de)serialization. This is on top of the low level classes which are already supported as part of the OpenSearch StreamInput/StreamOutput classes.

Any attempt to (de)serialize an unsupported class will lead to an exception.

Reply:
We won't need the safe class checks for Custom Serialization as we explicitly configure each of the classes eligible for (de)serialization. This is on top of the low level classes which are already supported as part of the OpenSearch StreamInput/StreamOutput classes.

Any attempt to (de)serialize an unsupported class will lead to an exception.

Reply:
We are tying two different concepts here: i) ability of a class to get serialized, ii) what classes are safe to be set in the header. Say, in future if a particular class need to blocked to be set in the header, it can be just excluded here but the class can be serialized and used otherwise.

Reply:
Got it. But it will not be possible to have the level of control as we had with JDK serialization. With custom serialization we use the StreamInput/StreamOutput classes which do not provide interfaces to restrict (de)serialization to certain list of classes.

It's guaranteed that anything that's not supported by StreamInput/StreamOutput + the Streamable and Writeable classes, will fail (de)serialization. If this is not OK, we will have to change extend the implementation of the StreamInput/StreamOutput classes in OpenSearch Core to provide this functionality.

Reply:
Can this logic moved up in the caller hierarchy, that way we are not loosing the functionality provided by the safe check?

Reply:
Analysed this a bit on my end. For the WRITEABLE and STREAMABLE serialization modes we can easily introduce the safe class checks.

For GENERIC mode we can override the `StreamOutput::writeGenericValue()` method to introduce safe class check. 

FYI the only classes which earlier were not allowed to be serialized but will be allowed as part of this change are -
```
ZonedDateTime
Text
ByteRef
ByteReference
Date
Object[] (an primitive generic array)
byte[]
```

Reply:
Have pushed changes for this.